torch/_refs/__init__.py

```diff
@@ -3804,86 +3916,89 @@ def _reshape_view_helper(a: TensorLikeType, *shape, allow_copy: bool) -> TensorL
+                # divisibility (accum % length != 0), and have deferred raising them, in the hope that we'd
+                # figure out a valid reshape later in the loop.
+                # But we failed, either by running out of dimensions, or we couldn't figure out the strides,
+                # and we've decided to re-raise to either graph break out, or provide the exact guard so the user
+                # can torch._check() to avoid this.
+                for f in deferred:
+                    f()
+
+            accum = a_.shape[idx]
+            end = idx
+            while True:
+                try:
+                    if accum % length == 0:
+                        break
+                except GuardOnDataDependentSymNode:
+                    deferred.append(lambda: bool(accum % length == 0))
+                if end == a_.ndim - 1:
+                    maybe_throw_dde()
+                end += 1
+                accum *= a_.shape[end]
+            if end != idx:
+                # NOTE: in this case multiple dimensions must be flatten to create the desired dimension
+                # This flattening is why reshape sometimes creates a copy -- because flattening
+                # may return a view of a copy
+
+                # Checks if collapse can be a view and short-circuits to copying reshape if it can't
+                new_shape, _new_strides = prims._collapse_view_helper(a_, idx, end)
+                if new_shape is None:
+                    if allow_copy:
+                        return prims.reshape(a, shape)
+
+                    maybe_throw_dde()
+                    msg = f"Cannot view a tensor with shape {a.shape} and strides {a.stride()} as a tensor with shape {shape}!"
+                    raise ValueError(msg)
+
+                a_ = flatten(a_, idx, end)
+
+            # Splits the (possibly flattened) dimension to create the desired dim length.
+            # guard_or_true is safe due to the tail unsqueeze routine.
+            if guard_or_true(accum != length):
+                a_ = prims.split_dim(a_, idx, length)
 
-        idx = idx + 1
+            idx = idx + 1
 
-    # Squeezes tail
-    while idx < a_.ndim:
-        torch._check(
-            a_.shape[idx] == 1,
-            lambda: f"a.size({idx}) expected to be 1 but got {a_.shape[idx]}",
-        )
-        a_ = squeeze(a_, idx)
+        # Squeezes tail
+        while idx < a_.ndim:
+            torch._check(
+                a_.shape[idx] == 1,
+                lambda: f"a.size({idx}) expected to be 1 but got {a_.shape[idx]}",
+            )
+            a_ = squeeze(a_, idx)
 
-    if a_ is a:
-        return prims.view_of(a)
-    else:
-        return a_
+        if a_ is a:
+            return prims.view_of(a)
+        else:
+            return a_
+    except GuardOnDataDependentSymNode as e:
+        return _unbacked_fallback_reshape_view(a, shape, allow_copy, e)
```

Comment:
I am not even if this function can ever be called during compile ! that said I will add manual testing for it for unbacked
(Call it explicitly maybe?)